File: backend/services/carbon_calculator.py
from services.emission_factors import (
    TRAVEL_FACTORS, FOOD_DIETS, FOOD_WASTE_PENALTY,
    ELECTRICITY_FACTOR, GAS_FACTOR, DEFAULT_CARBON_BUDGET,
    TRAVEL_KEY_LOOKUP, FOOD_KEY_LOOKUP
)


def calculate_travel_emissions(distance_km, transport_mode, passenger_count=1):
    """
    Calculate travel emissions in kg CO2.

    Args:
        distance_km: Distance traveled in kilometers.
        transport_mode: Display name like 'Petrol Car' or key like 'car_petrol'.
        passenger_count: Number of passengers (emissions divided equally).

    Returns:
        tuple: (emissions_kg, factor_used)
    """
    distance_km = float(distance_km)
    passenger_count = max(int(passenger_count), 1)

    mode_key = TRAVEL_KEY_LOOKUP.get(transport_mode, transport_mode)
    factor = TRAVEL_FACTORS.get(mode_key, 0.0)

    emissions = (distance_km * factor) / passenger_count
    return round(emissions, 2), factor


def calculate_food_emissions(diet_type, food_waste_enabled=False):
    """
    Calculate food emissions in kg CO2 per day.

    Args:
        diet_type: Display name like 'Omnivore' or key like 'omnivore'.
        food_waste_enabled: If True, applies 10% penalty.

    Returns:
        float: Emissions in kg CO2.
    """
    diet_key = FOOD_KEY_LOOKUP.get(diet_type, diet_type)
    base_emissions = FOOD_DIETS.get(diet_key, FOOD_DIETS["omnivore"])

    if food_waste_enabled:
        base_emissions *= FOOD_WASTE_PENALTY

    return round(base_emissions, 2)


def calculate_energy_emissions(electricity_kwh, gas_usage, renewable_percentage=0):
    """
    Calculate energy emissions in kg CO2.

    Args:
        electricity_kwh: Electricity consumed in kWh.
        gas_usage: Gas usage in kWh equivalent.
        renewable_percentage: Percentage of energy from renewables (0-100).

    Returns:
        float: Total energy emissions in kg CO2.
    """
    electricity_kwh = float(electricity_kwh)
    gas_usage = float(gas_usage)
    renewable_pct = max(0.0, min(100.0, float(renewable_percentage)))

    clean_modifier = 1.0 - (renewable_pct / 100.0)
    electricity_emissions = electricity_kwh * ELECTRICITY_FACTOR * clean_modifier
    gas_emissions = gas_usage * GAS_FACTOR

    return round(electricity_emissions + gas_emissions, 2)


def calculate_daily_total(travel_emissions, food_emissions, energy_emissions):
    """Sum category emissions into a daily total."""
    return round(
        float(travel_emissions) + float(food_emissions) + float(energy_emissions),
        2,
    )

File: frontend/src/utils/carbonCalculator.js
import {
  TRAVEL_FACTORS,
  TRAVEL_KEY_LOOKUP,
  FOOD_DIETS,
  FOOD_KEY_LOOKUP,
  FOOD_WASTE_PENALTY,
  ELECTRICITY_FACTOR,
  GAS_FACTOR,
} from "./emissionFactors";

/**
 * Calculate travel emissions in kg CO2.
 * @param {number} distanceKm - Distance in kilometers.
 * @param {string} transportMode - Display name or key.
 * @param {number} passengerCount - Number of passengers.
 * @returns {{ emissions: number, factor: number }}
 */
export function calculateTravelEmissions(distanceKm, transportMode, passengerCount = 1) {
  const distance = parseFloat(distanceKm) || 0;
  const passengers = Math.max(parseInt(passengerCount, 10) || 1, 1);
  const modeKey = TRAVEL_KEY_LOOKUP[transportMode] || transportMode;
  const factor = TRAVEL_FACTORS[modeKey] ?? 0;

  const emissions = (distance * factor) / passengers;
  return { emissions: round2(emissions), factor };
}

/**
 * Calculate food emissions in kg CO2 per day.
 * @param {string} dietType - Display name or key.
 * @param {boolean} foodWasteEnabled - Apply 10% penalty.
 * @returns {number} Emissions in kg CO2.
 */
export function calculateFoodEmissions(dietType, foodWasteEnabled = false) {
  const dietKey = FOOD_KEY_LOOKUP[dietType] || dietType;
  let base = FOOD_DIETS[dietKey] ?? FOOD_DIETS.omnivore;

  if (foodWasteEnabled) {
    base *= FOOD_WASTE_PENALTY;
  }

  return round2(base);
}

/**
 * Calculate energy emissions in kg CO2.
 * @param {number} electricityKwh
 * @param {number} gasUsage
 * @param {number} renewablePercentage - 0 to 100.
 * @returns {number}
 */
export function calculateEnergyEmissions(electricityKwh, gasUsage, renewablePercentage = 0) {
  const elec = parseFloat(electricityKwh) || 0;
  const gas = parseFloat(gasUsage) || 0;
  const renewable = Math.max(0, Math.min(100, parseFloat(renewablePercentage) || 0));

  const cleanModifier = 1 - renewable / 100;
  const elecEmissions = elec * ELECTRICITY_FACTOR * cleanModifier;
  const gasEmissions = gas * GAS_FACTOR;

  return round2(elecEmissions + gasEmissions);
}

/**
 * Sum category emissions into a daily total.
 */
export function calculateDailyTotal(travel = 0, food = 0, energy = 0) {
  return round2(parseFloat(travel) + parseFloat(food) + parseFloat(energy));
}

/**
 * Format a date string to a short label (MM/DD).
 */
export function formatDateLabel(dateStr) {
  const [, m, d] = dateStr.split("-");
  return `${m}/${d}`;
}

/**
 * Get heatmap level from emission value and budget.
 */
export function getHeatmapLevel(emission, budget) {
  if (emission === 0) return "none";
  if (emission <= budget * 0.7) return "low";
  if (emission <= budget) return "medium";
  return "high";
}

function round2(n) {
  return Math.round(n * 100) / 100;
}
